mkosi/initrd/__main__.py

```diff
@@ -0,0 +1,102 @@
+from mkosi.run import find_binary, run, uncaught_exception_handler
+from mkosi.types import PathString
+
+
+@uncaught_exception_handler()
+def main() -> None:
+    log_setup()
+
+    parser = argparse.ArgumentParser(
+        prog="mkosi-initrd",
+        description="Build initrds or Unified Kernel Images using mkosi",
+        allow_abbrev=False,
+        usage="mkosi-initrd [options...]",
+    )
+
+    parser.add_argument("--kernel-version",
+                        metavar="KERNEL_VERSION",
+                        help="Kernel version string",
+                        default=platform.uname().release)
+    parser.add_argument("--format",
+                        choices=["cpio", "uki"],
+                        help="Output format (CPIO archive or UKI)",
+                        default="cpio")
+    parser.add_argument("--version",
+                        action="version",
+                        version=f"mkosi {__version__}")
+
+    args_initrd, args_other = parser.parse_known_args()
+
+    cmdline: list[PathString] = [
+        "mkosi",
+        "--directory", "",
+        "--format", str(args_initrd.format),
+        "--workspace-dir=/var/tmp",
+        "--package-cache-dir=/var",
+        "--cache-only=metadata",
+        "--extra-tree", f"/usr/lib/modules/{args_initrd.kernel_version}:/usr/lib/modules/{args_initrd.kernel_version}",
+        "--extra-tree=/usr/lib/firmware:/usr/lib/firmware",
+        "--remove-files=/usr/lib/firmware/*-ucode",
+        "--kernel-modules-exclude=.*",
+        "--kernel-modules-include=host",
+        "--include=mkosi-initrd",
+    ]
+
+    for d in ("/usr/lib/mkosi-initrd", "/etc/mkosi-initrd"):
+        if Path(d).exists():
+            cmdline += ["--include", d]
+
+    with tempfile.TemporaryDirectory() as d:
+        # Make sure we don't use any of mkosi's default repositories.
+        for p in (
+            "yum.repos.d/mkosi.repo",
+            "apt/sources.list",
+            "zypp/repos.d/mkosi.repo",
+            "pacman.conf",
+        ):
+            (Path(d) / "etc" / p).parent.mkdir(parents=True, exist_ok=True)
+            (Path(d) / "etc" / p).touch()
+
+        # Copy in the host's package manager configuration.
+        for p in (
+            "dnf",
+            "yum.repos.d/",
+            "apt",
+            "zypp",
+            "pacman.conf",
+            "pacman.d/",
+        ):
+            if not (Path("/etc") / p).exists():
+                continue
+
+            (Path(d) / "etc" / p).parent.mkdir(parents=True, exist_ok=True)
+            if (Path("/etc") / p).resolve().is_file():
+                shutil.copy2(Path("/etc") / p, Path(d) / "etc" / p)
+            else:
+                shutil.copytree(Path("/etc") / p, Path(d) / "etc" / p,
+                                ignore=shutil.ignore_patterns("S.*"), dirs_exist_ok=True)
+
+        cmdline += ["--package-manager-tree", d]
+
+        if args_other is not None:
+            cmdline += args_other
```

Comment:
We should probably drop this, see above

Reply:
@aafeijoo-suse We definitely should not allow specifying arbitrary arguments. Some of them might make sense, but others definitely don't.

For example, specifying `--distribution` or `--release` is a very bad idea, because we disable usage of mkosi's builtin repositories and always use the ones from the host. So using anything other than the host's distribution and release in this case is a very bad idea.

This also makes me realize that we should very clearly document this tool is for building an initrd for the current host system and that for building generic initrds you need to invoke mkosi directly.

Reply:
ok, let's polish up the maybe-necessary missing arguments in later steps.